### PR TITLE
=act #16522 Optimize Stash creation

### PR DESCRIPTION
**Pull request by patriknw**
- cache the stash capacity config value

### CODE REVIEW

**Comment by patriknw** [comment]
JMH benchmark before:

    [info] Benchmark                              Mode     Cnt   Score   Error  Units
    [info] StashCreationBenchmark.testCreation  sample  343391  22.629 ± 6.159  us/opJMH benchmark before:

JMH benchmark after:

    [info] Benchmark                              Mode     Cnt   Score   Error  Units
    [info] StashCreationBenchmark.testCreation  sample  378135  18.695 ± 1.185  us/op

**Comment by patriknw** [comment]
Refs #16522

**Comment by ktoso** [comment]
LGTM, nice :-)

It reminded me of a fun piece of code in Http https://github.com/akka/akka/blob/release-2.3-dev/akka-http-core/src/main/scala/akka/http/impl/util/Rendering.scala#L53

File: akka-actor/src/main/scala/akka/actor/Stash.scala [hunk 1]
```diff
@@ -123,14 +123,8 @@ private[akka] trait StashSupport {
...
-  private val capacity: Int = {
-    val dispatcher = context.system.settings.config.getConfig(context.props.dispatcher)
-    val fallback = dispatcher.withFallback(context.system.settings.config.getConfig(Mailboxes.DefaultMailboxId))
-    val config =
-      if (context.props.mailbox == Mailboxes.DefaultMailboxId) fallback
-      else context.system.settings.config.getConfig(context.props.mailbox).withFallback(fallback)
-    config.getInt("stash-capacity")
-  }
+  private val capacity: Int =
+    context.system.mailboxes.stashCapacity(context.props.dispatcher, context.props.mailbox)
```
Comment (viktorklang) [inline akka-actor/src/main/scala/akka/actor/Stash.scala]: Ersteklasse

File: akka-actor/src/main/scala/akka/dispatch/Mailboxes.scala [hunk 2]
```diff
@@ -232,4 +232,38 @@ private[akka] class Mailboxes(
...
+
+  /**
+   * INTERNAL API: The capacity of the stash. Configured in the actor's mailbox or dispatcher config.
+   */
+  @tailrec private[akka] final def stashCapacity(dispatcher: String, mailbox: String): Int = {
+    if (dispatcher == Dispatchers.DefaultDispatcherId && mailbox == Mailboxes.DefaultMailboxId)
+      defaultStashCapacity
+    else {
+      val cache = stashCapacityCache.get
+      val key = dispatcher + "-" + mailbox
```
Comment (viktorklang) [inline akka-actor/src/main/scala/akka/dispatch/Mailboxes.scala]: No need to create this string for each recursive call

File: akka-actor/src/main/scala/akka/dispatch/Mailboxes.scala [hunk 3]
```diff
@@ -232,4 +232,38 @@ private[akka] class Mailboxes(
       .withFallback(settings.config.getConfig(id))
       .withFallback(defaultMailboxConfig)
   }
+
+  private var stashCapacityCache = new AtomicReference[Map[String, Int]](Map.empty[String, Int])
```
Comment (viktorklang) [inline akka-actor/src/main/scala/akka/dispatch/Mailboxes.scala]: a var to an Atomic reference?

Comment (patriknw) [inline akka-actor/src/main/scala/akka/dispatch/Mailboxes.scala]: 2x mutable :-) thanks

File: akka-actor/src/main/scala/akka/dispatch/Mailboxes.scala (hunk 2 above)
Comment (patriknw) [inline akka-actor/src/main/scala/akka/dispatch/Mailboxes.scala]: improved in f9ca7d1

**Comment by jrudolph** [comment]
+1 this is something I have seen in profiles.

**Comment by patriknw** [comment]
squashed, I'll merge after validation unless someone else want to review more

**Comment by viktorklang** [comment]
Everyone's had ample time to review this at this point and nobody's shouted they need extra time ;)

**Comment by patriknw** [comment]
(that was what I tried to say ;-))

**Comment by ktoso** [comment]
Looks Wunderbar To Me :wink: Merge at wil!

**Comment by viktorklang** [comment]
@patriknw merge? :)

**Pull request merged**